### PR TITLE
Port ubuntu vulnsrc to use OVAL rather than the Ubuntu CVE Tracker

### PR DESCRIPTION
**Pull request by alexmurray**
In response to issue #804

### CODE REVIEW

File: ext/vulnsrc/ubuntu/ubuntu.go [hunk 1]
```diff
@@ -13,369 +13,406 @@
...
+	return resp, nil
 }
 
-func (u *updater) Clean() {
-	if u.repositoryLocalPath != "" {
-		os.RemoveAll(u.repositoryLocalPath)
+// Get the latest modification time of a remote file
+// expressed as unix time
+func getLatestModifiedTime(url string) (int64, error) {
+	resp, err := http.Head(url)
```
Comment (ldelossa) [inline ext/vulnsrc/ubuntu/ubuntu.go:204]: @alexmurray  - believe you still want to defer close of the response body here. https://golang.org/pkg/net/http/ the docs groups the "head" method with the advice to close the body.

File: ext/vulnsrc/ubuntu/ubuntu.go [hunk 2]
```diff
@@ -13,369 +13,406 @@
...
-		"Patches": {},
-		// Product
-		"product": {},
-	}
+	ignoredCriterions          []string
+	ubuntuPackageCommentRegexp = regexp.MustCompile(`^(.*) package in ([a-z]+) (?:(?:was vulnerable|is related to the CVE in some way) but has been fixed \(note: '(.*)'\)|is affected and needs fixing).$`)
+	ubuntuOvalFileRegexp       = regexp.MustCompile(`com.ubuntu.([a-z]+).cve.oval.xml.bz2`)
+)
+
+type oval struct {
```
Comment (ldelossa) [inline ext/vulnsrc/ubuntu/ubuntu.go:61]: I don't have an issue with you defining the oval structure here, but you maybe interested in using our Oval parser fork: https://github.com/quay/goval-parser

**Review by ldelossa — CHANGES_REQUESTED** [review]
Looks good. one critique, one suggestion. 
May I assume you tested vulnerability matching in a test env of your own?

File: ext/vulnsrc/ubuntu/ubuntu.go (hunk 2 above)
Comment (alexmurray) [reply ext/vulnsrc/ubuntu/ubuntu.go:61]: I just used the example set by the existing OVAL vulnsrc's.

**Comment by alexmurray** [comment]
>  May I assume you tested vulnerability matching in a test env of your own?

Sadly I have not had any success at actually getting clair running properly locally - is there a simple quick-start-guide or similar to be able to deploy it on a local machine and scan a particular container to test?

File: ext/vulnsrc/ubuntu/ubuntu.go (hunk 1 above)
Comment (alexmurray) [reply ext/vulnsrc/ubuntu/ubuntu.go:204]: This was copied from the same function in the suse vulnsrc - if this is the case that should also be fixed too.

Comment (ldelossa) [reply ext/vulnsrc/ubuntu/ubuntu.go:204]: @alexmurray sounds like the right thing to do :+1:

Comment (alexmurray) [reply ext/vulnsrc/ubuntu/ubuntu.go:204]: Would  you prefer a separate PR for this or should I bundle it with this one?

Comment (ldelossa) [reply ext/vulnsrc/ubuntu/ubuntu.go:204]: Separate pr sounds good to me.

**Comment by ldelossa** [comment]
> > May I assume you tested vulnerability matching in a test env of your own?
> 
> Sadly I have not had any success at actually getting clair running properly locally - is there a simple quick-start-guide or similar to be able to deploy it on a local machine and scan a particular container to test?

@alexmurray this is fine. I will just need a bit to test out that everything works accordingly. 
I will also review the local development docs here: https://github.com/coreos/clair/blob/master/Documentation/local-development.md

**Comment by alexmurray** [comment]
Those docs assume contributors are highly familiar with kubernetes etc and already have this deployed - I am not familiar with k8s myself so perhaps a simple example using something like microk8s might be beneficial?

**Comment by alexmurray** [comment]
I have rebased this PR on top of current master to fix up the conflicts from the previous version - what can I do to help get this merged?

**Comment by cjwatson** [comment]
@ldelossa Please could you indicate whether there's anything else for @alexmurray to do here to satisfy your review comments?  While we've made some further changes to git.launchpad.net to make it easier to control how much this swamps our infrastructure, I'd really like to see this move along so that it stops being a problem.

**Comment by ldelossa** [comment]
@alexmurray @cjwatson my apologies for the delay. We have been putting all our engineering efforts into:
https://github.com/quay/claircore
and https://github.com/quay/clair/tree/development-4.0

I will merge this.

**Review by ldelossa — APPROVED** [review]

**Pull request merged**